On EveryBeat Block

sksum opened the pull request:
Added everybeatdonew in meter palette
this block opens up a new collapsed start block which contains a broadcast block. 
All the "action " duties are transferred to this new mouse. 
following snippet is a way we can use it 

![Screenshot (85)](https://user-images.githubusercontent.com/55449862/83629314-7e03d500-a5b7-11ea-88a8-488f3cc5aaa8.png)
 
currently , the beats per measure data is not transferred , everything is working in default bpm

walterbender commented:
This works really well.

I think it is OK that the start block appears on the screen. But I wonder about the mouse associated with the block. Maybe we need "hide mouse" and "show mouse" blocks to make the mouse disappear and/or reappear and add the "hide mouse" block to the macro?

We should coordinate with @meganindya re the new blocks as they will touch on his turtle.py refactoring.

walterbender commented:
Meanwhile, we should deprecate the current on-strong and on-weak beat blocks and have the new ones triggered by this block's dispatch rather than the on-every-note dispatch.

Some tricky bookkeeping though: what happens if you put the on-every-beat-do block in the trash? Presumably the associated start block needs to be trashed as well?

And if you have both an on-strong-beat block without an on-every-beat block then we need the start block, but we don't ever need more than one start block, regardless of how many on- blocks are used by a turtle.

Finally, what about BPM? We need to sync them up between the turtles. Not 100% sure how we should approach that. To start with, we should add a BPM block to the macro.

meganindya commented:
> This works really well.
> 
> I think it is OK that the start block appears on the screen. But I wonder about the mouse associated with the block. Maybe we need "hide mouse" and "show mouse" blocks to make the mouse disappear and/or reappear and add the "hide mouse" block to the macro?
> 
> We should coordinate with @meganindya re the new blocks as they will touch on his turtle.py refactoring.

`turtle.js` :p

pikurasa commented:
Walter is python programmer through and through

walterbender commented:
@sksum one other concern: the dispatch uses the turtle number in the name of the signal. But the turtle number might get changed when the project is reloaded. Maybe we need to assign a unique id to every turtle that is stored with the turtle and restored on load?

walterbender commented:
Or maybe the ID is actually tied to the on-every-beat-do block, not the turtle???

sksum commented:
yes the turtle id is linked to the block "on every beat do ",
maybe we can change the block color to match the new start block to avoid confusion

sksum commented:
Regarding BPM sync -
We can change this property inside the new mouse whenever the "on every beat block" is docked into a new start block.

pikurasa commented:
Testing now.

Maybe we move "on every note do" to Rhythm Palette (not meter)?

pikurasa commented:
It works well. Very fun.

One thing it does not yet seem to do is keep working after refresh.

```
Uncaught TypeError: Cannot read property 'listeners' of undefined
    at Logo._setListener (js/logo.js:5070)
    at EveryBeatDoBlockNew.flow (js/blocks/MeterBlocks.js:634)
    at Logo._runFromBlockNow (js/logo.js:1784)
    at js/logo.js:1483
```

sksum commented:
- [ ] BPM Sync
- [ ] Restoring links on Reload
- [ ] Trash Handling
- [ ] new On-beat and On-weak-beat blocks integration

pikurasa commented:
Question: What if the user does something somewhat contradictory such as the following?

(Meter assumes the beat is a quarter note, but the "beat turtle/mouse" is set to something else.)
![Screenshot at 2020-06-03 10:28:02 what-happens](https://user-images.githubusercontent.com/13454579/83649258-fb622200-a584-11ea-9e5b-a1d4eb030d54.png)

sksum commented:
maybe we can try to keep the "beat-turtle" always closed ,so that the user won't be able to change anything inside ?

pikurasa commented:
> maybe we can try to keep the "beat-turtle" always closed ,so that the user won't be able to change anything inside ?

The transparency is nice.

maybe a box block that says "beat" that gets its value from the meter block (if present)?

![Screenshot at 2020-06-03 10:38:11](https://user-images.githubusercontent.com/13454579/83650442-59433980-a586-11ea-81ba-eca66b7b1838.png)

pikurasa commented:
If we implement a "lock" feature as per https://github.com/sugarlabs/musicblocks/issues/1949 I would not mind "locking" the block by default. That way, one can see inside without messing anything up unintentionally.

sksum commented:
in this commit , I have implemented a slightly different approach :
A new BeatTurtle/Mouse will be added whenever we press play and deleted as soon as stopped.
, This way we wouldn't have to worry about trash handling , restoring or syncing meter every time we dock the block into a new turtle or change meter in existing turtle.

there still are some errors that need to be fixed though.  Thoughts ?

sksum commented:
I am getting an infinite loop while testing when reloading , I am working on it .
I am going to remove the review label  for now .

pikurasa commented:
It passes a very superficial check by me.

It may take me some time to get used to this new configuration, but it is nice to be able to work on the abstract level of beat.

Noteworthy: This works (I thought it might break)

![Screenshot at 2020-06-14 17:39:47 it works](https://user-images.githubusercontent.com/13454579/84604747-2fcbbd00-ae66-11ea-92b4-6a2c49033c8f.png)

Also, why does the broadcast block say "undefined" (i.e. `__everybeat_undefined__`)?

We need the help image as well.

walterbender commented:
Uncaught TypeError: turtles.getTurtleList is not a function
    at findTurtle (js/blocks/MeterBlocks.js:609)
    at EveryBeatDoBlockNew.flow (js/blocks/MeterBlocks.js:615)
    at Logo._runFromBlockNow (js/logo.js:1875)
    at js/logo.js:1544

And we need to think about a way to stop the beat turtle when the other turtle ends.

meganindya commented on js/blocks.js:
```diff
@@ -7066,6 +7089,13 @@ function Blocks(activity) {
             this.addDefaultBlock(parentBlock, thisBlock);
         }
 
+        if (myBlock.name == "everybeatdonew"){
+            let turtleID = myBlock.turtleID ;
+            let tur;
+            for(tur of turtles.getTurtleList())if (turtleID == tur.id)break;
+            this.sendStackToTrash(tur.startBlock);
+        }
```
There are a few leftover tasks in `Turtles`, however, `turtleList` has been updated to the new `getter` syntax, so access it like: `turtles.turleList`.

Also, I think it's better to use `{` `}` here.

walterbender commented:
I am thinking we may need to do the following:
(1) add a new block for "turtle id". It would be a hidden block and when it is created (internally, not from the palette), it would be assigned an id in much the same way that the signal name is constructed in your current code.
(2) add a new block for a boolean "turtle running". It would test to see if a turtle with an id (or name) is currently running.
(3) we can change the forever loop to a while "turtle running" <-- "turtle id"
(4) we can change the obscure signal name to "beat_" + "turtle id"

What do you think?

sksum commented:
> I am thinking we may need to do the following:
> (1) add a new block for "turtle id". It would be a hidden block and when it is created (internally, not from the palette), it would be assigned an id in much the same way that the signal name is constructed in your current code.
> (2) add a new block for a boolean "turtle running". It would test to see if a turtle with an id (or name) is currently running.
> (3) we can change the forever loop to a while "turtle running" <-- "turtle id"
> (4) we can change the obscure signal name to "beat_" + "turtle id"
> 
> What do you think?

looks good to me . 
will the boolean block be hidden too ?

walterbender commented:
I think the boolean block would be a useful addition to the Ensemble palette.
The reason to keep the id block off of the palette is because its value is calculated in the context of things like the on-beat-block. Not sure how that would work in a more general context.

walterbender commented:
I think I have a solution to the corner case I thought of yesterday.

If the "on every beat do" block is placed in an action as opposed to a start, then we don't yet know what turtle to associate it with. So we should not create the associated start stack until we first encounter the "on every beat do block" in execution of the code. At that time, we check to see if the turtle has a start stack associated with it -- if so, then proceed as usual. If not, create the stack on the fly. This way, multiple turtles can share an action and still get their own start stack associated with an on every beat do block. We'll have to start running the stack at that time as well, synched to the turtle it is associated with.

We may want to do it that way all the time -- to simplify the management of the BPM and meter -- which will be known at that time.

All of this in incumbent upon being able to create the blocks fast enough -- we could perhaps add the blocks to the blocklist, start the turtle, and only then worry about rendering the blocks on the screen. (If it is too slow, maybe we never render those blocks.)

walterbender commented:
As we discussed today, a "companion" field in turtle would allow us to keep the turtles in sync re BPM and meter.

pikurasa commented:
[On beat-Improv.html.zip](https://github.com/sugarlabs/musicblocks/files/4814723/On.beat-Improv.html.zip)
![Screenshot at 2020-06-22 13:25:54 on every beat improv](https://user-images.githubusercontent.com/13454579/85317554-88b1db80-b48c-11ea-9081-03296cfaa518.png)

A quick improvisation.

One observation: when the user first generates the beat block, the name of the start block appears as "beat", but using an ensemble block to "stop mouse = beat" does not stop it -- it cannot find the start block.

pikurasa commented:
@sksum when this lands, it may also help us close https://github.com/sugarlabs/musicblocks/issues/719

This is my sketch:
[Try-Beat-Factor-Functionality.html.zip](https://github.com/sugarlabs/musicblocks/files/4814816/Try-Beat-Factor-Functionality.html.zip)

The only problem with it at the moment is that the drums in action1 all play at once. I tried "no clock", but that does not seem to adjust the drums to their intended positions in time.

![Screenshot at 2020-06-22 13:51:33 on every beat and beat factor](https://user-images.githubusercontent.com/13454579/85319467-8a30d300-b48f-11ea-886f-823f8bfc788c.png)

walterbender commented:
Closing this PR as it is replaced by #2378
